packages/host/src/components/Initialise/hooks.ts

```diff
@@ -0,0 +1,130 @@
+import { useEffect, useState } from 'react';
+import create from 'zustand';
+import { AppRoute } from '@openmsupply-client/config';
+import {
+  AuthenticationError,
+  AuthError,
+  LocalStorage,
+  ServerStatus,
+  useNavigate,
+} from '@openmsupply-client/common';
+import { useHost } from '../../api/hooks';
+
+const SERVER_RESTART_TIMEOUT = 90000;
+const POLLING_INTERVAL = 3000;
+
+interface InitialiseForm {
+  error?: AuthenticationError;
+  isLoading: boolean;
+  password: string;
+  siteId?: number;
+  username: string;
+  url: string;
+  setError: (error?: AuthenticationError) => void;
+  setIsLoading: (isLoading: boolean) => void;
+  setPassword: (password: string) => void;
+  setUsername: (username: string) => void;
+  setSiteId: (siteId: number) => void;
+  setUrl: (url: string) => void;
+}
+
+const useInitialiseFormState = create<InitialiseForm>(set => ({
+  error: undefined,
+  isLoading: false,
+  password: '',
+  username: '',
+  url: '',
+  siteId: undefined,
+  setError: (error?: AuthenticationError) =>
+    set(state => ({ ...state, error })),
+  setIsLoading: (isLoading: boolean) => set(state => ({ ...state, isLoading })),
+  setPassword: (password: string) => set(state => ({ ...state, password })),
+  setUsername: (username: string) => set(state => ({ ...state, username })),
+  setUrl: (url: string) => set(state => ({ ...state, url })),
+  setSiteId: (siteId: number) => set(state => ({ ...state, siteId })),
+}));
+
+export const useInitialiseForm = () => {
+  const state = useInitialiseFormState();
+  const navigate = useNavigate();
+  const { mutateAsync: restart } = useHost.utils.restart();
+  const {
+    setIsLoading,
+    password,
+    setPassword,
+    username,
+    error,
+    setError,
+    url,
+    siteId,
+  } = state;
+  const [isPolling, setIsPolling] = useState(false);
+  const [isBootstrap, setIsBootstrap] = useState(false);
+  const { mutateAsync: update } = useHost.sync.update();
+  const { data } = useHost.utils.settings({
+    refetchInterval: POLLING_INTERVAL,
+    enabled: isPolling,
+  });
+
+  const onSave = async () => {
+    setError();
+    setIsLoading(true);
+    setIsBootstrap(false);
+    const syncSettings = {
+      centralServerSiteId: 1,
+      intervalSec: 300,
+      password,
+      siteHardwareId: '',
+      siteId: siteId || 2,
+      url,
+      username,
+    };
+
+    await update(syncSettings).catch(e => {
+      console.error(e);
+      setError({ message: 'Unable to save settings' });
+      setIsLoading(false);
+      return;
+    });
+    setPassword('');
+
+    await restart().catch(e => {
+      console.error(e);
+      setError({ message: 'Unable to restart the server' });
+      setIsLoading(false);
+      return;
+    });
+
+    setIsPolling(true);
+    LocalStorage.setItem('/auth/error', undefined);
```

Comment:
the query originally returns a valid response while shutting down; after a while it doesn't return anything, then returns an error and finally will return an authentication error, as the user isn't logged in.

The auth error is handled in GqlContext which sets an auth error on local storage - have set a listener for that and will wait for that response before redirecting. at this point the user can login.

With a demo datafile and running sqlite locally, it takes more than a minute for the whole process - so have set 90s as the timeout

Reply:
Ok maybe we should have public /status endpoint containing general information, e.g. api version and server status so that it can be polled without permissions?

I guess for sync status permissions would be required?

Reply:
an open endpoint which returns status & version would be great - we could remove some of the hackyness here 😄 
and I think is safe enough to have publicly available. Should be no need to have sync status public that I can see, though some aspect of it as part of the status would be needed - the client needs to know if the user can log in, which requires sync to have completed.

So either

```
{
  version: String
  status: ServerStatus
  sync: Active | Idle
}
```

or

```
export enum ServerStatus {
  Running = 'RUNNING',
  /** Server misses configuration to start up fully */
  Stage_0 = 'STAGE_0',
  Synchronising = 'SYNCHRONISING'
}
```

or 🤷